### PR TITLE
Return size_t from shape::size().

### PR DESCRIPTION
**Pull request by ZhennanQin**
This issue blocks MXNet PR: https://github.com/apache/incubator-mxnet/pull/13697. https://github.com/apache/incubator-mxnet/pull/13729 is the corresponding fix for MXNet. After this fix, that PR can pass. Ref to https://github.com/apache/incubator-mxnet/pull/13755.

### CODE REVIEW

**Comment by ZhennanQin** [comment]
@apeforest Please help review and merge. Thanks.

**Comment by pengzhao-intel** [comment]
@szha We have an offline talk with @apeforest and we agreed with the change to fix the casting issue in MXNet. Please help take a review and merge this PR.

Happy New Year 👍

**Review by apeforest — COMMENTED** [review]
@ZhennanQin Could you please link the PR (ready for review) in MXNet that requires this change?

**Comment by KellenSunderland** [comment]
It's https://github.com/apache/incubator-mxnet/pull/13697  If we cast manually to size_t it compiles.  For reference: https://github.com/apache/incubator-mxnet/pull/13755

**Comment by ZhennanQin** [comment]
@szha Please help to review and merge this. Then I will upgrade mshadow  in https://github.com/apache/incubator-mxnet/pull/13729.

File: mshadow/tensor.h [hunk 1]
```diff
@@ -123,7 +123,7 @@ struct Shape {
     return s;
   }
   /*! \return number of valid elements */
-  MSHADOW_XINLINE index_t Size(void) const {
+  MSHADOW_XINLINE size_t Size(void) const {
     index_t size = this->shape_[0];
```
Comment (apeforest) [inline mshadow/tensor.h:127]: Maybe we want to cast here to size_t explicitly?

**Comment by stu1130** [comment]
@ZhennanQin could you address the comments Thanks!

**Comment by ZhennanQin** [comment]
@stu1130 The comment request is beyond my initial scope, and I don't have enough knowledge to address other API changes. As it's not a block issue for me anymore, let @apeforest to continue this work. I can abandon this PR.

**Pull request closed without merging**